rust/worker/src/execution/operators/materialize_logs.rs

```diff
@@ -0,0 +1,142 @@
+use crate::execution::operator::Operator;
+use crate::segment::record_segment::RecordSegmentReaderCreationError;
+use crate::segment::{materialize_logs, record_segment::RecordSegmentReader};
+use crate::segment::{LogMaterializerError, MaterializedLogRecord};
+use async_trait::async_trait;
+use chroma_blockstore::provider::BlockfileProvider;
+use chroma_error::ChromaError;
+use chroma_types::{Chunk, LogRecord, Segment};
+use futures::{FutureExt, TryFutureExt};
+use ouroboros::self_referencing;
+use std::sync::atomic::AtomicU32;
+use std::sync::Arc;
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum MaterializeLogOperatorError {
+    #[error("Could not create record segment reader: {0}")]
+    RecordSegmentReaderCreationFailed(#[from] RecordSegmentReaderCreationError),
+    #[error("Log materialization failed: {0}")]
+    LogMaterializationFailed(#[from] LogMaterializerError),
+}
+
+impl ChromaError for MaterializeLogOperatorError {
+    fn code(&self) -> chroma_error::ErrorCodes {
+        match self {
+            MaterializeLogOperatorError::RecordSegmentReaderCreationFailed(e) => e.code(),
+            MaterializeLogOperatorError::LogMaterializationFailed(e) => e.code(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct MaterializeLogOperator {}
+
+impl MaterializeLogOperator {
+    pub fn new() -> Box<Self> {
+        Box::new(MaterializeLogOperator {})
+    }
+}
+
+#[derive(Debug)]
+pub struct MaterializeLogInput {
+    logs: Chunk<LogRecord>,
+    provider: BlockfileProvider,
+    record_segment: Segment,
+    offset_id: Arc<AtomicU32>,
+}
+
+impl MaterializeLogInput {
+    pub fn new(
+        logs: Chunk<LogRecord>,
+        provider: BlockfileProvider,
+        record_segment: Segment,
+        offset_id: Arc<AtomicU32>,
+    ) -> Self {
+        MaterializeLogInput {
+            logs,
+            provider,
+            record_segment,
+            offset_id,
+        }
+    }
+}
+
+#[self_referencing(no_doc)]
```

Comment:
✨ the magic part

Reply:
I'd like an overview of how this magic works before we adopt it.

Reply:
I might end up implementing this differently with pin project (will not change anything in PRs up the stack)
cc @rescrv